components/prism-bash.js

```diff
@@ -19,66 +29,168 @@
 					punctuation: /\(\(?|\)\)?|,|;/
 				}
 			},
-			// Command Substitution
+			// [1]: Command Substitution
 			{
 				pattern: /\$\([^)]+\)|`[^`]+`/,
 				greedy: true,
 				inside: {
 					variable: /^\$\(|^`|\)$|`$/
 				}
 			},
-			/\$(?:[\w#?*!@]+|\{[^}]+\})/i
-		]
+            // [2]: Brace expansion
+            {
+                pattern: /\$(?:\{[^}]+\})/i,
+                greedy: true,
+                inside: {
+                    operator: /:[-=?+]?|[!\/]|\#\#?|%%?|\^\^?|,,?/,
+                    punctuation: /[\[\]]/,
+					environment: {
+						pattern: new RegExp("(\\{)(?:" + envVarList + ")(?=$|\\}|\\W)"),
+						lookbehind: true,
+						alias: 'constant'
+					}
+                }
+            },
+			/\$(?:\w+|#|\?|\*|!|@|\$)/i
+		],
+        // man echo / printf, + \"
+        entity: /\\([abceEfnrtv\\"]|O?[0-7]{1,3}|x[0-9a-fA-F]{1,2}|u[0-9a-fA-F]{4}|U[0-9a-fA-F]{8})/
 	};
 
 	Prism.languages.bash = {
-		'shebang': {
-			pattern: /^#!\s*\/bin\/bash|^#!\s*\/bin\/sh/,
+		shebang: {
+			pattern: /^#!\s*\/.*/,
 			alias: 'important'
 		},
-		'comment': {
-			pattern: /(^|[^"{\\])#.*/,
+		comment: {
+			pattern: /(^|[^"{\\$])#.*/,
+			lookbehind: true
+		},
+		'function-name': [
+			// a) function foo {
+			// b) foo() {
+			// c) function foo() {
+			// but not “foo {”
+			{
+				// a) and c)
+				pattern: /((^|\W)function\s+)(?:\w+)(?=(\s*\(\s*\))?\s*\{)/,
+				lookbehind: true
+			},
+			{
+				// b)
+				pattern: /(^|\W)(?:\w+)(?=\s*\(\s*\)\s*\{)/,
+				lookbehind: true
+			}
+		],
+		// Highlight variable names as variables in for and select beginnings.
+		'for-or-select': {
+			pattern: /((^|[\s;|&]|[<>]\()(for|select)\s+)(?:\w+)(?=\s+in\s)/,
+			alias: 'variable',
+			lookbehind: true
+		},
+		// Highlight variable names as variables in the left-hand part
+		// of assignments (“=” and “+=”).
+		'assign-left': {
+			pattern: /(^|[\s;|&]|[<>]\()\w+(?=\+?=)/,
+			inside: {
+				environment: {
+					pattern: new RegExp("(^|[\\s;|&]|[<>]\\()(?:" + envVarList + ")(?=$|\\W)"),
+					lookbehind: true,
+					alias: 'constant'
+				}
+			},
+			alias: 'variable',
 			lookbehind: true
 		},
-		'string': [
-			//Support for Here-Documents https://en.wikipedia.org/wiki/Here_document
+		string: [
+			// Support for Here-documents https://en.wikipedia.org/wiki/Here_document
 			{
-				pattern: /((?:^|[^<])<<\s*)["']?(\w+?)["']?\s*\r?\n(?:[\s\S])*?\r?\n\2/,
+				pattern: /((?:^|[^<])<<-?\s*)(\w+?)\s*\r?\n(?:[\s\S])*?\r?\n\2/,
```

Comment:
Isn't this group used to make sure the here document is ended with the same tag? Like, to prevent:

```
cat << _PLOP_
blabla
_PLAP_
```

(`_PLOP_` ≠ `_PLAP_`)
There's a `\2` for this at the end.
The tests go haywire if I ungroup this thing.

Reply:
Woops you're right! I missed the `\2` there. Sorry about that!